File: 3. DFSBFS/특정 거리의 도시 찾기.py
from collections import deque

def bfs():
  q = deque()
  q.append(start)
  distance[start] = 0
  
  while q:
    now = q.popleft()

    for next in graph[now]:
      if distance[next] == INF:
        distance[next] = distance[now] + 1
        if distance[next] == target:
          ans.append(next)
        q.append(next)

v, e, target, start = map(int, input().split())
INF = int(1e9)
distance = [INF] * (v+1)
graph = [[] for _ in range(v+1)]

for _ in range(e):
  a, b = map(int, input().split())
  graph[a].append(b)

ans = []
bfs()
ans.sort()
if not len(ans):
  print(-1)
else:
  for i in ans:
    print(i)

File: 3. DFSBFS/특정 거리의 도시 찾기.js
const data =
`
1 2
1 3
2 3
2 4
`.trim().split('\n')

const [v, e, target, start] = [4, 4, 2, 1]

const INF = 1e9
const distance = new Array(v + 1).fill(INF)
const graph = new Array(v + 1).fill().map(() => [])

for (i of data) {
  const[a, b] = i.split(' ').map(x => parseInt(x))
  graph[a].push(b)
}
const ans = []
bfs()
ans.sort((a, b) => (a-b))
if (ans.length === 0) {
  console.log(-1)
}
else {
  for (i of ans) {
    console.log(i)
  }
}

function bfs() {
  q = []
  q.push(start)
  distance[start] = 0

  while (q.length != 0) {
    const now = q.shift()
    
    for (next of graph[now]) {
      if (distance[next] === INF) {
        distance[next] = distance[now] + 1
        if (distance[next] === target) {
          ans.push(next)
        }
        q.push(next)
      }
    }
  }
}